vector/src/main/java/im/vector/riotx/features/home/ShortcutsHandler.kt

```diff
@@ -0,0 +1,87 @@
+ * limitations under the License.
+ */
+
+package im.vector.riotx.features.home
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.graphics.Bitmap
+import android.os.Build
+import androidx.core.content.pm.ShortcutInfoCompat
+import androidx.core.content.pm.ShortcutManagerCompat
+import androidx.core.graphics.drawable.IconCompat
+import im.vector.matrix.android.api.session.room.model.tag.RoomTag
+import im.vector.matrix.android.api.util.toMatrixItem
+import im.vector.riotx.core.glide.GlideApp
+import im.vector.riotx.core.utils.DimensionConverter
+import im.vector.riotx.features.home.room.detail.RoomDetailActivity
+import io.reactivex.schedulers.Schedulers
+import javax.inject.Inject
+
+private val useAdaptiveIcon = Build.VERSION.SDK_INT >= Build.VERSION_CODES.O
+private const val adaptiveIconSizeDp = 108
+private const val adaptiveIconOuterSidesDp = 18
+
+class ShortcutsHandler @Inject constructor(
+        private val homeRoomListStore: HomeRoomListDataSource,
+        private val avatarRenderer: AvatarRenderer,
+        private val dimensionConverter: DimensionConverter
+) {
+
+    @SuppressLint("CheckResult")
+    fun observeRoomsAndBuildShortcuts(context: Context) {
+        homeRoomListStore
+                .observe()
+                .observeOn(Schedulers.computation())
+                .subscribe { rooms ->
+                    val shortcuts = rooms
+                            .filter { room -> room.tags.any { it.name == RoomTag.ROOM_TAG_FAVOURITE } }
+                            .take(n = 4) // Android only allows us to create 4 shortcuts
+                            .map { room ->
+                                val intent = RoomDetailActivity.shortcutIntent(context, room.roomId)
+
+                                val adaptiveIconSize = dimensionConverter.dpToPx(adaptiveIconSizeDp)
+                                val adaptiveIconOuterSides = dimensionConverter.dpToPx(adaptiveIconOuterSidesDp)
+                                val size = if (useAdaptiveIcon) {
+                                    adaptiveIconSize - adaptiveIconOuterSides
+                                } else {
+                                    dimensionConverter.dpToPx(72)
+                                }
+
+                                val bitmap = avatarRenderer.shortcutDrawable(context, GlideApp.with(context), room.toMatrixItem(), size)
+
+                                ShortcutInfoCompat.Builder(context, room.roomId)
+                                        .setShortLabel(room.displayName)
+                                        .setIcon(bitmap.toProfileImageIcon())
+                                        .setIntent(intent)
+                                        .build()
+                            }
+
+                    ShortcutManagerCompat.removeAllDynamicShortcuts(context)
+                    ShortcutManagerCompat.addDynamicShortcuts(context, shortcuts)
```

Comment:
The shrotcuts do not need to be updated if the list is the same. You should add a distinct() call somewhere, either in th Rx stream, or by keeping a memory copy of the current shortcut. This API is rate limited, because it can consume system resource (just read that from the doc)